src/buf.c

```diff
@@ -266,4 +285,10 @@ struct Buf *buf_append (struct Buf *buf, const void *ptr, int n_elem)
 	return buf;
 }
 
+struct Buf *buf_append_alloc (struct Buf *buf, const void *ptr, int n_elem)
+{
+	buf_append(&alloc_buf, ptr, n_elem);
```

Comment:
whydoubt
> While Buf is somewhat like a std::vector, Buf's interface only allows for appending entries, not for removing or changing them.

> I was careful to limit alloc_buf use to buf.c, so the scope is actually file (though I did forget to set it static).
> This actually seems like a very normal pattern if you think of buf_init_alloc as initializing the buf interface
> buf_destroy_alloc as finalizing.

A class-wide alloc_buf is a bad idea when what you are trying to fix only matters to a particular _instantiation_ of a _generic_ type. In this case, I mean the `Buf` of allocated strings.

> What you suggest will involve either altering the Buf structure to contain its own list of free-able elements,
or require every element be free-able (come from a strdup, etc.). In the former case, there would still need to be buf_append_alloc. In the latter case, we would either need change Buf to know about the element type (currently it just knows about the element size) maybe with an enum, or there would need to be a separate variant of buf_destroy() and if you want the interface to take care of the strdup then we still need buf_append_alloc.

Actually, I was expecting the latter. If a Buf instance contains an array of strings, then either all of its strings shall be free-able or none at all. If a Buf instance are strings that should be free'd when destroyed, then do this subroutine:

```
void buf_strings_destroy (struct Buf *buf)
{
	char *p;
	if (!buf) return;
	for (p = buf->elts + (buf->nelts - 1) * buf->elt_size; p >= buf->elts; p -= buf->elt_size) {
		free(*p);
	}
	free(buf->elts);
	buf->elts = NULL;
}
```

You may consider this as subclassing---I just create a variant of the finalizer that calls free() on all of its members as well as freeing the array itself.